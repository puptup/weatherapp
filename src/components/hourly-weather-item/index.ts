export { HourlyWeatherItem } from "./hourly-weather-item";
