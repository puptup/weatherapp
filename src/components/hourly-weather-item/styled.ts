import styled from "styled-components";

export const Image = styled.img`
  display: block;
  margin: 0 auto;
  width: 40px;
  height: 40px;
`;

export const Item = styled.div`
  width: 60px;
  background-color: ${({ theme }) => theme.palette.item.primary};
  border-radius: ${({ theme }) => theme.spacing.borderRadiusItem};
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  padding: ${({ theme }) => theme.spacing.padding};
  margin: ${({ theme }) => theme.spacing.blockOutside};
  flex-shrink: 0;
`;
