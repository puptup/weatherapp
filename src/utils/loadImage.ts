const comparer = {
  "01d": 2,
  "01n": 26,
  "02d": 3,
  "02n": 27,
  "03d": 7,
  "03n": 7,
  "04d": 5,
  "04n": 29,
  "09d": 13,
  "09n": 32,
  "10d": 11,
  "10n": 11,
  "11d": 14,
  "11n": 14,
  "13d": 19,
  "13n": 34,
  "50d": 9,
  "50n": 9,
};

type ComapereKeys = keyof typeof comparer;

const getOpenWeatherIcon = (iconID: string) => {
  const newIconID = comparer[iconID as ComapereKeys];
  return newIconID || 1;
};

export const loadIcon = async (iconID: string, openWeather: boolean = false) => {
  if (openWeather) {
    return import(`../assets/weather-icons/${getOpenWeatherIcon(iconID)}.png`);
  }
  return import(`../assets/weather-icons/${iconID}.png`);
};
