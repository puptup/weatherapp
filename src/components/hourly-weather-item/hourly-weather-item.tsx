import { hoursOptions } from "@constants/date-and-time";
import { HourlyWeather } from "@types";
import { loadIcon } from "@utils/loadImage";
import React, { useEffect, useState } from "react";

import { Image, Item } from "./styled";

type HourlyWeatherItemProps = {
  item: HourlyWeather;
};

export const HourlyWeatherItem = ({ item }: HourlyWeatherItemProps) => {
  const { date, temperature, icon } = item;
  const [image, setImage] = useState<string>();

  useEffect(() => {
    loadIcon(icon, true).then((path) => {
      setImage(path.default);
    });
  }, []);

  return (
    <Item>
      <div>{new Date(date).toLocaleTimeString("en-GB", hoursOptions)}</div>
      <Image src={image} alt="hourly weahter" />
      <div>{Math.round(temperature)}</div>
    </Item>
  );
};
