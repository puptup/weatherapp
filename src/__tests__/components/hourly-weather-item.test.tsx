import "@testing-library/jest-dom";

import { HourlyWeatherItem } from "@components/hourly-weather-item/hourly-weather-item";
import * as utils from "@utils/loadImage";
import { renderWithProviders } from "@utils/testing/render-with-providers";
import React from "react";

describe("hourly-weather-item", () => {
  jest.spyOn(utils, "loadIcon").mockReturnValue(Promise.resolve(() => "test"));

  it("should correspond to snapshot", () => {
    const { getByTestId } = renderWithProviders(
      <HourlyWeatherItem
        item={{
          icon: "test",
          date: "2023-01-23T14:13:22.224Z",
          maxTemperature: 0,
          minTemperature: 0,
          temperature: 0,
          weather: "test",
        }}
      />
    );
    expect(getByTestId("hourly-weather-item")).toMatchSnapshot();
  });
});
