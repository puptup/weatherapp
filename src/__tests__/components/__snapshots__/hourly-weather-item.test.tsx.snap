// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hourly-weather-item should correspond to snapshot 1`] = `
<div
  class="sc-gswNZR bsuyEe"
  data-testid="hourly-weather-item"
>
  <div>
    17
  </div>
  <img
    alt="hourly weahter"
    class="sc-bcXHqe bzmlG"
  />
  <div>
    0
  </div>
</div>
`;
